// @polymeshassociation/polymesh-sdk/polkadot/nft/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/naming-convention */
exports.default = {
    rpc: {
        validateNFTTransfer: {
            description: 'Verifies if and the sender and receiver are not the same, if both have valid balances, if the sender owns the nft, and if all compliance rules are being respected.',
            params: [
                {
                    name: 'sender_portfolio',
                    type: 'PortfolioId',
                    isOptional: false,
                },
                {
                    name: 'receiver_portfolio',
                    type: 'PortfolioId',
                    isOptional: false,
                },
                {
                    name: 'nfts',
                    type: 'NFTs',
                    isOptional: false,
                },
                {
                    name: 'blockHash',
                    type: 'Hash',
                    isOptional: true,
                },
            ],
            type: 'DispatchResult',
        },
    },
    types: {},
};
