// @polymeshassociation/polymesh-sdk/api/procedures/toggleFreezeTransfers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toggleFreezeTransfers = exports.getAuthorization = exports.prepareToggleFreezeTransfers = void 0;
const internal_1 = require("../../internal");
const types_1 = require("../../types");
const conversion_1 = require("../../utils/conversion");
/**
 * @hidden
 */
function prepareToggleFreezeTransfers(args) {
    return __awaiter(this, void 0, void 0, function* () {
        const { context: { polymeshApi: { tx: { asset }, }, }, context, } = this;
        const { ticker, freeze } = args;
        const rawTicker = (0, conversion_1.stringToTicker)(ticker, context);
        const assetEntity = new internal_1.BaseAsset({ ticker }, context);
        const isFrozen = yield assetEntity.isFrozen();
        if (freeze) {
            if (isFrozen) {
                throw new internal_1.PolymeshError({
                    code: types_1.ErrorCode.NoDataChange,
                    message: 'The Asset is already frozen',
                });
            }
            return {
                transaction: asset.freeze,
                args: [rawTicker],
                resolver: undefined,
            };
        }
        if (!isFrozen) {
            throw new internal_1.PolymeshError({
                code: types_1.ErrorCode.NoDataChange,
                message: 'The Asset is already unfrozen',
            });
        }
        return {
            transaction: asset.unfreeze,
            args: [rawTicker],
            resolver: undefined,
        };
    });
}
exports.prepareToggleFreezeTransfers = prepareToggleFreezeTransfers;
/**
 * @hidden
 */
function getAuthorization({ ticker, freeze }) {
    return {
        permissions: {
            transactions: [freeze ? types_1.TxTags.asset.Freeze : types_1.TxTags.asset.Unfreeze],
            assets: [new internal_1.BaseAsset({ ticker }, this.context)],
            portfolios: [],
        },
    };
}
exports.getAuthorization = getAuthorization;
/**
 * @hidden
 */
const toggleFreezeTransfers = () => new internal_1.Procedure(prepareToggleFreezeTransfers, getAuthorization);
exports.toggleFreezeTransfers = toggleFreezeTransfers;
