// @polymeshassociation/polymesh-sdk/api/procedures/togglePauseRequirements.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.togglePauseRequirements = exports.getAuthorization = exports.prepareTogglePauseRequirements = void 0;
const internal_1 = require("../../internal");
const types_1 = require("../../types");
const conversion_1 = require("../../utils/conversion");
/**
 * @hidden
 */
function prepareTogglePauseRequirements(args) {
    return __awaiter(this, void 0, void 0, function* () {
        const { context: { polymeshApi: { query, tx }, }, context, } = this;
        const { ticker, pause } = args;
        const rawTicker = (0, conversion_1.stringToTicker)(ticker, context);
        const { paused } = yield query.complianceManager.assetCompliances(rawTicker);
        if (pause === (0, conversion_1.boolToBoolean)(paused)) {
            throw new internal_1.PolymeshError({
                code: types_1.ErrorCode.NoDataChange,
                message: `Requirements are already ${paused ? '' : 'un'}paused`,
            });
        }
        return {
            transaction: pause
                ? tx.complianceManager.pauseAssetCompliance
                : tx.complianceManager.resumeAssetCompliance,
            args: [rawTicker],
            resolver: undefined,
        };
    });
}
exports.prepareTogglePauseRequirements = prepareTogglePauseRequirements;
/**
 * @hidden
 */
function getAuthorization({ ticker, pause }) {
    return {
        permissions: {
            transactions: [
                pause
                    ? types_1.TxTags.complianceManager.PauseAssetCompliance
                    : types_1.TxTags.complianceManager.ResumeAssetCompliance,
            ],
            assets: [new internal_1.FungibleAsset({ ticker }, this.context)],
            portfolios: [],
        },
    };
}
exports.getAuthorization = getAuthorization;
/**
 * @hidden
 */
const togglePauseRequirements = () => new internal_1.Procedure(prepareTogglePauseRequirements, getAuthorization);
exports.togglePauseRequirements = togglePauseRequirements;
