// @polymeshassociation/polymesh-sdk/api/procedures/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllowanceOperation = exports.InstructionAffirmationOperation = exports.ClaimOperation = exports.TransferRestrictionType = void 0;
var TransferRestrictionType;
(function (TransferRestrictionType) {
    TransferRestrictionType["Count"] = "Count";
    TransferRestrictionType["Percentage"] = "Percentage";
    TransferRestrictionType["ClaimCount"] = "ClaimCount";
    TransferRestrictionType["ClaimPercentage"] = "ClaimPercentage";
})(TransferRestrictionType = exports.TransferRestrictionType || (exports.TransferRestrictionType = {}));
var ClaimOperation;
(function (ClaimOperation) {
    ClaimOperation["Revoke"] = "Revoke";
    ClaimOperation["Add"] = "Add";
    ClaimOperation["Edit"] = "Edit";
})(ClaimOperation = exports.ClaimOperation || (exports.ClaimOperation = {}));
var InstructionAffirmationOperation;
(function (InstructionAffirmationOperation) {
    InstructionAffirmationOperation["Affirm"] = "Affirm";
    InstructionAffirmationOperation["Withdraw"] = "Withdraw";
    InstructionAffirmationOperation["Reject"] = "Reject";
})(InstructionAffirmationOperation = exports.InstructionAffirmationOperation || (exports.InstructionAffirmationOperation = {}));
var AllowanceOperation;
(function (AllowanceOperation) {
    AllowanceOperation["Set"] = "Set";
    AllowanceOperation["Increase"] = "Increase";
    AllowanceOperation["Decrease"] = "Decrease";
})(AllowanceOperation = exports.AllowanceOperation || (exports.AllowanceOperation = {}));
