// @polymeshassociation/polymesh-sdk/api/entities/Instruction/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AffirmationStatus = exports.InstructionType = exports.InstructionStatus = void 0;
var InstructionStatus;
(function (InstructionStatus) {
    InstructionStatus["Pending"] = "Pending";
    InstructionStatus["Failed"] = "Failed";
    InstructionStatus["Success"] = "Success";
    InstructionStatus["Rejected"] = "Rejected";
})(InstructionStatus = exports.InstructionStatus || (exports.InstructionStatus = {}));
var InstructionType;
(function (InstructionType) {
    InstructionType["SettleOnAffirmation"] = "SettleOnAffirmation";
    InstructionType["SettleOnBlock"] = "SettleOnBlock";
    InstructionType["SettleManual"] = "SettleManual";
})(InstructionType = exports.InstructionType || (exports.InstructionType = {}));
var AffirmationStatus;
(function (AffirmationStatus) {
    AffirmationStatus["Unknown"] = "Unknown";
    AffirmationStatus["Pending"] = "Pending";
    AffirmationStatus["Affirmed"] = "Affirmed";
})(AffirmationStatus = exports.AffirmationStatus || (exports.AffirmationStatus = {}));
