// @polymeshassociation/polymesh-sdk/generated/types.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxTags = exports.ModuleName = exports.NftTx = exports.ContractsTx = exports.PreimageTx = exports.PolymeshContractsTx = exports.TestUtilsTx = exports.RewardsTx = exports.RelayerTx = exports.ExternalAgentsTx = exports.UtilityTx = exports.TreasuryTx = exports.StoTx = exports.StatisticsTx = exports.SettlementTx = exports.SchedulerTx = exports.ProtocolFeeTx = exports.PortfolioTx = exports.PipsTx = exports.CorporateBallotTx = exports.CorporateActionTx = exports.ComplianceManagerTx = exports.CheckpointTx = exports.CapitalDistributionTx = exports.AssetTx = exports.SudoTx = exports.ImOnlineTx = exports.GrandpaTx = exports.SessionTx = exports.StakingTx = exports.BridgeTx = exports.MultiSigTx = exports.UpgradeCommitteeMembershipTx = exports.UpgradeCommitteeTx = exports.TechnicalCommitteeMembershipTx = exports.TechnicalCommitteeTx = exports.CommitteeMembershipTx = exports.PolymeshCommitteeTx = exports.CddServiceProvidersTx = exports.IdentityTx = exports.BalancesTx = exports.AuthorshipTx = exports.IndicesTx = exports.TimestampTx = exports.BabeTx = exports.SystemTx = exports.CountryCode = void 0;
var CountryCode;
(function (CountryCode) {
    CountryCode["Af"] = "Af";
    CountryCode["Ax"] = "Ax";
    CountryCode["Al"] = "Al";
    CountryCode["Dz"] = "Dz";
    CountryCode["As"] = "As";
    CountryCode["Ad"] = "Ad";
    CountryCode["Ao"] = "Ao";
    CountryCode["Ai"] = "Ai";
    CountryCode["Aq"] = "Aq";
    CountryCode["Ag"] = "Ag";
    CountryCode["Ar"] = "Ar";
    CountryCode["Am"] = "Am";
    CountryCode["Aw"] = "Aw";
    CountryCode["Au"] = "Au";
    CountryCode["At"] = "At";
    CountryCode["Az"] = "Az";
    CountryCode["Bs"] = "Bs";
    CountryCode["Bh"] = "Bh";
    CountryCode["Bd"] = "Bd";
    CountryCode["Bb"] = "Bb";
    CountryCode["By"] = "By";
    CountryCode["Be"] = "Be";
    CountryCode["Bz"] = "Bz";
    CountryCode["Bj"] = "Bj";
    CountryCode["Bm"] = "Bm";
    CountryCode["Bt"] = "Bt";
    CountryCode["Bo"] = "Bo";
    CountryCode["Ba"] = "Ba";
    CountryCode["Bw"] = "Bw";
    CountryCode["Bv"] = "Bv";
    CountryCode["Br"] = "Br";
    CountryCode["Vg"] = "Vg";
    CountryCode["Io"] = "Io";
    CountryCode["Bn"] = "Bn";
    CountryCode["Bg"] = "Bg";
    CountryCode["Bf"] = "Bf";
    CountryCode["Bi"] = "Bi";
    CountryCode["Kh"] = "Kh";
    CountryCode["Cm"] = "Cm";
    CountryCode["Ca"] = "Ca";
    CountryCode["Cv"] = "Cv";
    CountryCode["Ky"] = "Ky";
    CountryCode["Cf"] = "Cf";
    CountryCode["Td"] = "Td";
    CountryCode["Cl"] = "Cl";
    CountryCode["Cn"] = "Cn";
    CountryCode["Hk"] = "Hk";
    CountryCode["Mo"] = "Mo";
    CountryCode["Cx"] = "Cx";
    CountryCode["Cc"] = "Cc";
    CountryCode["Co"] = "Co";
    CountryCode["Km"] = "Km";
    CountryCode["Cg"] = "Cg";
    CountryCode["Cd"] = "Cd";
    CountryCode["Ck"] = "Ck";
    CountryCode["Cr"] = "Cr";
    CountryCode["Ci"] = "Ci";
    CountryCode["Hr"] = "Hr";
    CountryCode["Cu"] = "Cu";
    CountryCode["Cy"] = "Cy";
    CountryCode["Cz"] = "Cz";
    CountryCode["Dk"] = "Dk";
    CountryCode["Dj"] = "Dj";
    CountryCode["Dm"] = "Dm";
    CountryCode["Do"] = "Do";
    CountryCode["Ec"] = "Ec";
    CountryCode["Eg"] = "Eg";
    CountryCode["Sv"] = "Sv";
    CountryCode["Gq"] = "Gq";
    CountryCode["Er"] = "Er";
    CountryCode["Ee"] = "Ee";
    CountryCode["Et"] = "Et";
    CountryCode["Fk"] = "Fk";
    CountryCode["Fo"] = "Fo";
    CountryCode["Fj"] = "Fj";
    CountryCode["Fi"] = "Fi";
    CountryCode["Fr"] = "Fr";
    CountryCode["Gf"] = "Gf";
    CountryCode["Pf"] = "Pf";
    CountryCode["Tf"] = "Tf";
    CountryCode["Ga"] = "Ga";
    CountryCode["Gm"] = "Gm";
    CountryCode["Ge"] = "Ge";
    CountryCode["De"] = "De";
    CountryCode["Gh"] = "Gh";
    CountryCode["Gi"] = "Gi";
    CountryCode["Gr"] = "Gr";
    CountryCode["Gl"] = "Gl";
    CountryCode["Gd"] = "Gd";
    CountryCode["Gp"] = "Gp";
    CountryCode["Gu"] = "Gu";
    CountryCode["Gt"] = "Gt";
    CountryCode["Gg"] = "Gg";
    CountryCode["Gn"] = "Gn";
    CountryCode["Gw"] = "Gw";
    CountryCode["Gy"] = "Gy";
    CountryCode["Ht"] = "Ht";
    CountryCode["Hm"] = "Hm";
    CountryCode["Va"] = "Va";
    CountryCode["Hn"] = "Hn";
    CountryCode["Hu"] = "Hu";
    CountryCode["Is"] = "Is";
    CountryCode["In"] = "In";
    CountryCode["Id"] = "Id";
    CountryCode["Ir"] = "Ir";
    CountryCode["Iq"] = "Iq";
    CountryCode["Ie"] = "Ie";
    CountryCode["Im"] = "Im";
    CountryCode["Il"] = "Il";
    CountryCode["It"] = "It";
    CountryCode["Jm"] = "Jm";
    CountryCode["Jp"] = "Jp";
    CountryCode["Je"] = "Je";
    CountryCode["Jo"] = "Jo";
    CountryCode["Kz"] = "Kz";
    CountryCode["Ke"] = "Ke";
    CountryCode["Ki"] = "Ki";
    CountryCode["Kp"] = "Kp";
    CountryCode["Kr"] = "Kr";
    CountryCode["Kw"] = "Kw";
    CountryCode["Kg"] = "Kg";
    CountryCode["La"] = "La";
    CountryCode["Lv"] = "Lv";
    CountryCode["Lb"] = "Lb";
    CountryCode["Ls"] = "Ls";
    CountryCode["Lr"] = "Lr";
    CountryCode["Ly"] = "Ly";
    CountryCode["Li"] = "Li";
    CountryCode["Lt"] = "Lt";
    CountryCode["Lu"] = "Lu";
    CountryCode["Mk"] = "Mk";
    CountryCode["Mg"] = "Mg";
    CountryCode["Mw"] = "Mw";
    CountryCode["My"] = "My";
    CountryCode["Mv"] = "Mv";
    CountryCode["Ml"] = "Ml";
    CountryCode["Mt"] = "Mt";
    CountryCode["Mh"] = "Mh";
    CountryCode["Mq"] = "Mq";
    CountryCode["Mr"] = "Mr";
    CountryCode["Mu"] = "Mu";
    CountryCode["Yt"] = "Yt";
    CountryCode["Mx"] = "Mx";
    CountryCode["Fm"] = "Fm";
    CountryCode["Md"] = "Md";
    CountryCode["Mc"] = "Mc";
    CountryCode["Mn"] = "Mn";
    CountryCode["Me"] = "Me";
    CountryCode["Ms"] = "Ms";
    CountryCode["Ma"] = "Ma";
    CountryCode["Mz"] = "Mz";
    CountryCode["Mm"] = "Mm";
    CountryCode["Na"] = "Na";
    CountryCode["Nr"] = "Nr";
    CountryCode["Np"] = "Np";
    CountryCode["Nl"] = "Nl";
    CountryCode["An"] = "An";
    CountryCode["Nc"] = "Nc";
    CountryCode["Nz"] = "Nz";
    CountryCode["Ni"] = "Ni";
    CountryCode["Ne"] = "Ne";
    CountryCode["Ng"] = "Ng";
    CountryCode["Nu"] = "Nu";
    CountryCode["Nf"] = "Nf";
    CountryCode["Mp"] = "Mp";
    CountryCode["No"] = "No";
    CountryCode["Om"] = "Om";
    CountryCode["Pk"] = "Pk";
    CountryCode["Pw"] = "Pw";
    CountryCode["Ps"] = "Ps";
    CountryCode["Pa"] = "Pa";
    CountryCode["Pg"] = "Pg";
    CountryCode["Py"] = "Py";
    CountryCode["Pe"] = "Pe";
    CountryCode["Ph"] = "Ph";
    CountryCode["Pn"] = "Pn";
    CountryCode["Pl"] = "Pl";
    CountryCode["Pt"] = "Pt";
    CountryCode["Pr"] = "Pr";
    CountryCode["Qa"] = "Qa";
    CountryCode["Re"] = "Re";
    CountryCode["Ro"] = "Ro";
    CountryCode["Ru"] = "Ru";
    CountryCode["Rw"] = "Rw";
    CountryCode["Bl"] = "Bl";
    CountryCode["Sh"] = "Sh";
    CountryCode["Kn"] = "Kn";
    CountryCode["Lc"] = "Lc";
    CountryCode["Mf"] = "Mf";
    CountryCode["Pm"] = "Pm";
    CountryCode["Vc"] = "Vc";
    CountryCode["Ws"] = "Ws";
    CountryCode["Sm"] = "Sm";
    CountryCode["St"] = "St";
    CountryCode["Sa"] = "Sa";
    CountryCode["Sn"] = "Sn";
    CountryCode["Rs"] = "Rs";
    CountryCode["Sc"] = "Sc";
    CountryCode["Sl"] = "Sl";
    CountryCode["Sg"] = "Sg";
    CountryCode["Sk"] = "Sk";
    CountryCode["Si"] = "Si";
    CountryCode["Sb"] = "Sb";
    CountryCode["So"] = "So";
    CountryCode["Za"] = "Za";
    CountryCode["Gs"] = "Gs";
    CountryCode["Ss"] = "Ss";
    CountryCode["Es"] = "Es";
    CountryCode["Lk"] = "Lk";
    CountryCode["Sd"] = "Sd";
    CountryCode["Sr"] = "Sr";
    CountryCode["Sj"] = "Sj";
    CountryCode["Sz"] = "Sz";
    CountryCode["Se"] = "Se";
    CountryCode["Ch"] = "Ch";
    CountryCode["Sy"] = "Sy";
    CountryCode["Tw"] = "Tw";
    CountryCode["Tj"] = "Tj";
    CountryCode["Tz"] = "Tz";
    CountryCode["Th"] = "Th";
    CountryCode["Tl"] = "Tl";
    CountryCode["Tg"] = "Tg";
    CountryCode["Tk"] = "Tk";
    CountryCode["To"] = "To";
    CountryCode["Tt"] = "Tt";
    CountryCode["Tn"] = "Tn";
    CountryCode["Tr"] = "Tr";
    CountryCode["Tm"] = "Tm";
    CountryCode["Tc"] = "Tc";
    CountryCode["Tv"] = "Tv";
    CountryCode["Ug"] = "Ug";
    CountryCode["Ua"] = "Ua";
    CountryCode["Ae"] = "Ae";
    CountryCode["Gb"] = "Gb";
    CountryCode["Us"] = "Us";
    CountryCode["Um"] = "Um";
    CountryCode["Uy"] = "Uy";
    CountryCode["Uz"] = "Uz";
    CountryCode["Vu"] = "Vu";
    CountryCode["Ve"] = "Ve";
    CountryCode["Vn"] = "Vn";
    CountryCode["Vi"] = "Vi";
    CountryCode["Wf"] = "Wf";
    CountryCode["Eh"] = "Eh";
    CountryCode["Ye"] = "Ye";
    CountryCode["Zm"] = "Zm";
    CountryCode["Zw"] = "Zw";
    CountryCode["Bq"] = "Bq";
    CountryCode["Cw"] = "Cw";
    CountryCode["Sx"] = "Sx";
})(CountryCode = exports.CountryCode || (exports.CountryCode = {}));
var SystemTx;
(function (SystemTx) {
    SystemTx["FillBlock"] = "system.fillBlock";
    SystemTx["Remark"] = "system.remark";
    SystemTx["SetHeapPages"] = "system.setHeapPages";
    SystemTx["SetCode"] = "system.setCode";
    SystemTx["SetCodeWithoutChecks"] = "system.setCodeWithoutChecks";
    SystemTx["SetChangesTrieConfig"] = "system.setChangesTrieConfig";
    SystemTx["SetStorage"] = "system.setStorage";
    SystemTx["KillStorage"] = "system.killStorage";
    SystemTx["KillPrefix"] = "system.killPrefix";
    SystemTx["RemarkWithEvent"] = "system.remarkWithEvent";
    SystemTx["PlaceholderFillBlock"] = "system.placeholderFillBlock";
})(SystemTx = exports.SystemTx || (exports.SystemTx = {}));
var BabeTx;
(function (BabeTx) {
    BabeTx["ReportEquivocation"] = "babe.reportEquivocation";
    BabeTx["ReportEquivocationUnsigned"] = "babe.reportEquivocationUnsigned";
    BabeTx["PlanConfigChange"] = "babe.planConfigChange";
})(BabeTx = exports.BabeTx || (exports.BabeTx = {}));
var TimestampTx;
(function (TimestampTx) {
    TimestampTx["Set"] = "timestamp.set";
})(TimestampTx = exports.TimestampTx || (exports.TimestampTx = {}));
var IndicesTx;
(function (IndicesTx) {
    IndicesTx["Claim"] = "indices.claim";
    IndicesTx["Transfer"] = "indices.transfer";
    IndicesTx["Free"] = "indices.free";
    IndicesTx["ForceTransfer"] = "indices.forceTransfer";
    IndicesTx["Freeze"] = "indices.freeze";
})(IndicesTx = exports.IndicesTx || (exports.IndicesTx = {}));
var AuthorshipTx;
(function (AuthorshipTx) {
    AuthorshipTx["SetUncles"] = "authorship.setUncles";
})(AuthorshipTx = exports.AuthorshipTx || (exports.AuthorshipTx = {}));
var BalancesTx;
(function (BalancesTx) {
    BalancesTx["Transfer"] = "balances.transfer";
    BalancesTx["TransferWithMemo"] = "balances.transferWithMemo";
    BalancesTx["DepositBlockRewardReserveBalance"] = "balances.depositBlockRewardReserveBalance";
    BalancesTx["SetBalance"] = "balances.setBalance";
    BalancesTx["ForceTransfer"] = "balances.forceTransfer";
    BalancesTx["BurnAccountBalance"] = "balances.burnAccountBalance";
})(BalancesTx = exports.BalancesTx || (exports.BalancesTx = {}));
var IdentityTx;
(function (IdentityTx) {
    IdentityTx["CddRegisterDid"] = "identity.cddRegisterDid";
    IdentityTx["InvalidateCddClaims"] = "identity.invalidateCddClaims";
    IdentityTx["RemoveSecondaryKeys"] = "identity.removeSecondaryKeys";
    IdentityTx["AcceptPrimaryKey"] = "identity.acceptPrimaryKey";
    IdentityTx["ChangeCddRequirementForMkRotation"] = "identity.changeCddRequirementForMkRotation";
    IdentityTx["JoinIdentityAsKey"] = "identity.joinIdentityAsKey";
    IdentityTx["LeaveIdentityAsKey"] = "identity.leaveIdentityAsKey";
    IdentityTx["AddClaim"] = "identity.addClaim";
    IdentityTx["RevokeClaim"] = "identity.revokeClaim";
    IdentityTx["SetPermissionToSigner"] = "identity.setPermissionToSigner";
    IdentityTx["LegacySetPermissionToSigner"] = "identity.legacySetPermissionToSigner";
    IdentityTx["FreezeSecondaryKeys"] = "identity.freezeSecondaryKeys";
    IdentityTx["UnfreezeSecondaryKeys"] = "identity.unfreezeSecondaryKeys";
    IdentityTx["AddAuthorization"] = "identity.addAuthorization";
    IdentityTx["RemoveAuthorization"] = "identity.removeAuthorization";
    IdentityTx["AddSecondaryKeysWithAuthorization"] = "identity.addSecondaryKeysWithAuthorization";
    IdentityTx["AddInvestorUniquenessClaim"] = "identity.addInvestorUniquenessClaim";
    IdentityTx["GcAddCddClaim"] = "identity.gcAddCddClaim";
    IdentityTx["GcRevokeCddClaim"] = "identity.gcRevokeCddClaim";
    IdentityTx["AddInvestorUniquenessClaimV2"] = "identity.addInvestorUniquenessClaimV2";
    IdentityTx["RevokeClaimByIndex"] = "identity.revokeClaimByIndex";
    IdentityTx["RotatePrimaryKeyToSecondary"] = "identity.rotatePrimaryKeyToSecondary";
    IdentityTx["RemoveSecondaryKeysOld"] = "identity.removeSecondaryKeysOld";
    IdentityTx["PlaceholderLegacySetPermissionToSigner"] = "identity.placeholderLegacySetPermissionToSigner";
    IdentityTx["AddSecondaryKeysWithAuthorizationOld"] = "identity.addSecondaryKeysWithAuthorizationOld";
    IdentityTx["SetSecondaryKeyPermissions"] = "identity.setSecondaryKeyPermissions";
    IdentityTx["RegisterCustomClaimType"] = "identity.registerCustomClaimType";
    IdentityTx["CddRegisterDidWithCdd"] = "identity.cddRegisterDidWithCdd";
    IdentityTx["CreateChildIdentity"] = "identity.createChildIdentity";
    IdentityTx["CreateChildIdentities"] = "identity.createChildIdentities";
    IdentityTx["UnlinkChildIdentity"] = "identity.unlinkChildIdentity";
})(IdentityTx = exports.IdentityTx || (exports.IdentityTx = {}));
var CddServiceProvidersTx;
(function (CddServiceProvidersTx) {
    CddServiceProvidersTx["SetActiveMembersLimit"] = "cddServiceProviders.setActiveMembersLimit";
    CddServiceProvidersTx["DisableMember"] = "cddServiceProviders.disableMember";
    CddServiceProvidersTx["AddMember"] = "cddServiceProviders.addMember";
    CddServiceProvidersTx["RemoveMember"] = "cddServiceProviders.removeMember";
    CddServiceProvidersTx["SwapMember"] = "cddServiceProviders.swapMember";
    CddServiceProvidersTx["ResetMembers"] = "cddServiceProviders.resetMembers";
    CddServiceProvidersTx["AbdicateMembership"] = "cddServiceProviders.abdicateMembership";
})(CddServiceProvidersTx = exports.CddServiceProvidersTx || (exports.CddServiceProvidersTx = {}));
var PolymeshCommitteeTx;
(function (PolymeshCommitteeTx) {
    PolymeshCommitteeTx["SetVoteThreshold"] = "polymeshCommittee.setVoteThreshold";
    PolymeshCommitteeTx["SetReleaseCoordinator"] = "polymeshCommittee.setReleaseCoordinator";
    PolymeshCommitteeTx["SetExpiresAfter"] = "polymeshCommittee.setExpiresAfter";
    PolymeshCommitteeTx["VoteOrPropose"] = "polymeshCommittee.voteOrPropose";
    PolymeshCommitteeTx["Vote"] = "polymeshCommittee.vote";
})(PolymeshCommitteeTx = exports.PolymeshCommitteeTx || (exports.PolymeshCommitteeTx = {}));
var CommitteeMembershipTx;
(function (CommitteeMembershipTx) {
    CommitteeMembershipTx["SetActiveMembersLimit"] = "committeeMembership.setActiveMembersLimit";
    CommitteeMembershipTx["DisableMember"] = "committeeMembership.disableMember";
    CommitteeMembershipTx["AddMember"] = "committeeMembership.addMember";
    CommitteeMembershipTx["RemoveMember"] = "committeeMembership.removeMember";
    CommitteeMembershipTx["SwapMember"] = "committeeMembership.swapMember";
    CommitteeMembershipTx["ResetMembers"] = "committeeMembership.resetMembers";
    CommitteeMembershipTx["AbdicateMembership"] = "committeeMembership.abdicateMembership";
})(CommitteeMembershipTx = exports.CommitteeMembershipTx || (exports.CommitteeMembershipTx = {}));
var TechnicalCommitteeTx;
(function (TechnicalCommitteeTx) {
    TechnicalCommitteeTx["SetVoteThreshold"] = "technicalCommittee.setVoteThreshold";
    TechnicalCommitteeTx["SetReleaseCoordinator"] = "technicalCommittee.setReleaseCoordinator";
    TechnicalCommitteeTx["SetExpiresAfter"] = "technicalCommittee.setExpiresAfter";
    TechnicalCommitteeTx["VoteOrPropose"] = "technicalCommittee.voteOrPropose";
    TechnicalCommitteeTx["Vote"] = "technicalCommittee.vote";
})(TechnicalCommitteeTx = exports.TechnicalCommitteeTx || (exports.TechnicalCommitteeTx = {}));
var TechnicalCommitteeMembershipTx;
(function (TechnicalCommitteeMembershipTx) {
    TechnicalCommitteeMembershipTx["SetActiveMembersLimit"] = "technicalCommitteeMembership.setActiveMembersLimit";
    TechnicalCommitteeMembershipTx["DisableMember"] = "technicalCommitteeMembership.disableMember";
    TechnicalCommitteeMembershipTx["AddMember"] = "technicalCommitteeMembership.addMember";
    TechnicalCommitteeMembershipTx["RemoveMember"] = "technicalCommitteeMembership.removeMember";
    TechnicalCommitteeMembershipTx["SwapMember"] = "technicalCommitteeMembership.swapMember";
    TechnicalCommitteeMembershipTx["ResetMembers"] = "technicalCommitteeMembership.resetMembers";
    TechnicalCommitteeMembershipTx["AbdicateMembership"] = "technicalCommitteeMembership.abdicateMembership";
})(TechnicalCommitteeMembershipTx = exports.TechnicalCommitteeMembershipTx || (exports.TechnicalCommitteeMembershipTx = {}));
var UpgradeCommitteeTx;
(function (UpgradeCommitteeTx) {
    UpgradeCommitteeTx["SetVoteThreshold"] = "upgradeCommittee.setVoteThreshold";
    UpgradeCommitteeTx["SetReleaseCoordinator"] = "upgradeCommittee.setReleaseCoordinator";
    UpgradeCommitteeTx["SetExpiresAfter"] = "upgradeCommittee.setExpiresAfter";
    UpgradeCommitteeTx["VoteOrPropose"] = "upgradeCommittee.voteOrPropose";
    UpgradeCommitteeTx["Vote"] = "upgradeCommittee.vote";
})(UpgradeCommitteeTx = exports.UpgradeCommitteeTx || (exports.UpgradeCommitteeTx = {}));
var UpgradeCommitteeMembershipTx;
(function (UpgradeCommitteeMembershipTx) {
    UpgradeCommitteeMembershipTx["SetActiveMembersLimit"] = "upgradeCommitteeMembership.setActiveMembersLimit";
    UpgradeCommitteeMembershipTx["DisableMember"] = "upgradeCommitteeMembership.disableMember";
    UpgradeCommitteeMembershipTx["AddMember"] = "upgradeCommitteeMembership.addMember";
    UpgradeCommitteeMembershipTx["RemoveMember"] = "upgradeCommitteeMembership.removeMember";
    UpgradeCommitteeMembershipTx["SwapMember"] = "upgradeCommitteeMembership.swapMember";
    UpgradeCommitteeMembershipTx["ResetMembers"] = "upgradeCommitteeMembership.resetMembers";
    UpgradeCommitteeMembershipTx["AbdicateMembership"] = "upgradeCommitteeMembership.abdicateMembership";
})(UpgradeCommitteeMembershipTx = exports.UpgradeCommitteeMembershipTx || (exports.UpgradeCommitteeMembershipTx = {}));
var MultiSigTx;
(function (MultiSigTx) {
    MultiSigTx["CreateMultisig"] = "multiSig.createMultisig";
    MultiSigTx["CreateOrApproveProposalAsIdentity"] = "multiSig.createOrApproveProposalAsIdentity";
    MultiSigTx["CreateOrApproveProposalAsKey"] = "multiSig.createOrApproveProposalAsKey";
    MultiSigTx["CreateProposalAsIdentity"] = "multiSig.createProposalAsIdentity";
    MultiSigTx["CreateProposalAsKey"] = "multiSig.createProposalAsKey";
    MultiSigTx["ApproveAsIdentity"] = "multiSig.approveAsIdentity";
    MultiSigTx["ApproveAsKey"] = "multiSig.approveAsKey";
    MultiSigTx["RejectAsIdentity"] = "multiSig.rejectAsIdentity";
    MultiSigTx["RejectAsKey"] = "multiSig.rejectAsKey";
    MultiSigTx["AcceptMultisigSignerAsIdentity"] = "multiSig.acceptMultisigSignerAsIdentity";
    MultiSigTx["AcceptMultisigSignerAsKey"] = "multiSig.acceptMultisigSignerAsKey";
    MultiSigTx["AddMultisigSigner"] = "multiSig.addMultisigSigner";
    MultiSigTx["RemoveMultisigSigner"] = "multiSig.removeMultisigSigner";
    MultiSigTx["AddMultisigSignersViaCreator"] = "multiSig.addMultisigSignersViaCreator";
    MultiSigTx["RemoveMultisigSignersViaCreator"] = "multiSig.removeMultisigSignersViaCreator";
    MultiSigTx["ChangeSigsRequired"] = "multiSig.changeSigsRequired";
    MultiSigTx["MakeMultisigSigner"] = "multiSig.makeMultisigSigner";
    MultiSigTx["MakeMultisigPrimary"] = "multiSig.makeMultisigPrimary";
    MultiSigTx["ExecuteScheduledProposal"] = "multiSig.executeScheduledProposal";
    MultiSigTx["MakeMultisigSecondary"] = "multiSig.makeMultisigSecondary";
    MultiSigTx["ChangeSigsRequiredViaCreator"] = "multiSig.changeSigsRequiredViaCreator";
    MultiSigTx["RemoveCreatorControls"] = "multiSig.removeCreatorControls";
})(MultiSigTx = exports.MultiSigTx || (exports.MultiSigTx = {}));
var BridgeTx;
(function (BridgeTx) {
    BridgeTx["ChangeController"] = "bridge.changeController";
    BridgeTx["ChangeAdmin"] = "bridge.changeAdmin";
    BridgeTx["ChangeTimelock"] = "bridge.changeTimelock";
    BridgeTx["Freeze"] = "bridge.freeze";
    BridgeTx["Unfreeze"] = "bridge.unfreeze";
    BridgeTx["ChangeBridgeLimit"] = "bridge.changeBridgeLimit";
    BridgeTx["ChangeBridgeExempted"] = "bridge.changeBridgeExempted";
    BridgeTx["ForceHandleBridgeTx"] = "bridge.forceHandleBridgeTx";
    BridgeTx["BatchProposeBridgeTx"] = "bridge.batchProposeBridgeTx";
    BridgeTx["ProposeBridgeTx"] = "bridge.proposeBridgeTx";
    BridgeTx["HandleBridgeTx"] = "bridge.handleBridgeTx";
    BridgeTx["FreezeTxs"] = "bridge.freezeTxs";
    BridgeTx["UnfreezeTxs"] = "bridge.unfreezeTxs";
    BridgeTx["HandleScheduledBridgeTx"] = "bridge.handleScheduledBridgeTx";
    BridgeTx["AddFreezeAdmin"] = "bridge.addFreezeAdmin";
    BridgeTx["RemoveFreezeAdmin"] = "bridge.removeFreezeAdmin";
    BridgeTx["RemoveTxs"] = "bridge.removeTxs";
})(BridgeTx = exports.BridgeTx || (exports.BridgeTx = {}));
var StakingTx;
(function (StakingTx) {
    StakingTx["Bond"] = "staking.bond";
    StakingTx["BondExtra"] = "staking.bondExtra";
    StakingTx["Unbond"] = "staking.unbond";
    StakingTx["WithdrawUnbonded"] = "staking.withdrawUnbonded";
    StakingTx["Validate"] = "staking.validate";
    StakingTx["Nominate"] = "staking.nominate";
    StakingTx["Chill"] = "staking.chill";
    StakingTx["SetPayee"] = "staking.setPayee";
    StakingTx["SetController"] = "staking.setController";
    StakingTx["SetValidatorCount"] = "staking.setValidatorCount";
    StakingTx["IncreaseValidatorCount"] = "staking.increaseValidatorCount";
    StakingTx["ScaleValidatorCount"] = "staking.scaleValidatorCount";
    StakingTx["AddPermissionedValidator"] = "staking.addPermissionedValidator";
    StakingTx["RemovePermissionedValidator"] = "staking.removePermissionedValidator";
    StakingTx["ValidateCddExpiryNominators"] = "staking.validateCddExpiryNominators";
    StakingTx["SetCommissionCap"] = "staking.setCommissionCap";
    StakingTx["SetMinBondThreshold"] = "staking.setMinBondThreshold";
    StakingTx["ForceNoEras"] = "staking.forceNoEras";
    StakingTx["ForceNewEra"] = "staking.forceNewEra";
    StakingTx["SetInvulnerables"] = "staking.setInvulnerables";
    StakingTx["ForceUnstake"] = "staking.forceUnstake";
    StakingTx["ForceNewEraAlways"] = "staking.forceNewEraAlways";
    StakingTx["CancelDeferredSlash"] = "staking.cancelDeferredSlash";
    StakingTx["PayoutStakers"] = "staking.payoutStakers";
    StakingTx["Rebond"] = "staking.rebond";
    StakingTx["SetHistoryDepth"] = "staking.setHistoryDepth";
    StakingTx["ReapStash"] = "staking.reapStash";
    StakingTx["SubmitElectionSolution"] = "staking.submitElectionSolution";
    StakingTx["SubmitElectionSolutionUnsigned"] = "staking.submitElectionSolutionUnsigned";
    StakingTx["PayoutStakersBySystem"] = "staking.payoutStakersBySystem";
    StakingTx["ChangeSlashingAllowedFor"] = "staking.changeSlashingAllowedFor";
    StakingTx["UpdatePermissionedValidatorIntendedCount"] = "staking.updatePermissionedValidatorIntendedCount";
    StakingTx["ChillFromGovernance"] = "staking.chillFromGovernance";
})(StakingTx = exports.StakingTx || (exports.StakingTx = {}));
var SessionTx;
(function (SessionTx) {
    SessionTx["SetKeys"] = "session.setKeys";
    SessionTx["PurgeKeys"] = "session.purgeKeys";
})(SessionTx = exports.SessionTx || (exports.SessionTx = {}));
var GrandpaTx;
(function (GrandpaTx) {
    GrandpaTx["ReportEquivocation"] = "grandpa.reportEquivocation";
    GrandpaTx["ReportEquivocationUnsigned"] = "grandpa.reportEquivocationUnsigned";
    GrandpaTx["NoteStalled"] = "grandpa.noteStalled";
})(GrandpaTx = exports.GrandpaTx || (exports.GrandpaTx = {}));
var ImOnlineTx;
(function (ImOnlineTx) {
    ImOnlineTx["Heartbeat"] = "imOnline.heartbeat";
})(ImOnlineTx = exports.ImOnlineTx || (exports.ImOnlineTx = {}));
var SudoTx;
(function (SudoTx) {
    SudoTx["Sudo"] = "sudo.sudo";
    SudoTx["SudoUncheckedWeight"] = "sudo.sudoUncheckedWeight";
    SudoTx["SetKey"] = "sudo.setKey";
    SudoTx["SudoAs"] = "sudo.sudoAs";
})(SudoTx = exports.SudoTx || (exports.SudoTx = {}));
var AssetTx;
(function (AssetTx) {
    AssetTx["RegisterTicker"] = "asset.registerTicker";
    AssetTx["AcceptTickerTransfer"] = "asset.acceptTickerTransfer";
    AssetTx["AcceptAssetOwnershipTransfer"] = "asset.acceptAssetOwnershipTransfer";
    AssetTx["CreateAsset"] = "asset.createAsset";
    AssetTx["Freeze"] = "asset.freeze";
    AssetTx["Unfreeze"] = "asset.unfreeze";
    AssetTx["RenameAsset"] = "asset.renameAsset";
    AssetTx["Issue"] = "asset.issue";
    AssetTx["Redeem"] = "asset.redeem";
    AssetTx["MakeDivisible"] = "asset.makeDivisible";
    AssetTx["AddDocuments"] = "asset.addDocuments";
    AssetTx["RemoveDocuments"] = "asset.removeDocuments";
    AssetTx["SetFundingRound"] = "asset.setFundingRound";
    AssetTx["UpdateIdentifiers"] = "asset.updateIdentifiers";
    AssetTx["ClaimClassicTicker"] = "asset.claimClassicTicker";
    AssetTx["ReserveClassicTicker"] = "asset.reserveClassicTicker";
    AssetTx["ControllerTransfer"] = "asset.controllerTransfer";
    AssetTx["RegisterCustomAssetType"] = "asset.registerCustomAssetType";
    AssetTx["CreateAssetWithCustomType"] = "asset.createAssetWithCustomType";
    AssetTx["SetAssetMetadata"] = "asset.setAssetMetadata";
    AssetTx["SetAssetMetadataDetails"] = "asset.setAssetMetadataDetails";
    AssetTx["RegisterAndSetLocalAssetMetadata"] = "asset.registerAndSetLocalAssetMetadata";
    AssetTx["RegisterAssetMetadataLocalType"] = "asset.registerAssetMetadataLocalType";
    AssetTx["RegisterAssetMetadataGlobalType"] = "asset.registerAssetMetadataGlobalType";
    AssetTx["RedeemFromPortfolio"] = "asset.redeemFromPortfolio";
    AssetTx["UpdateAssetType"] = "asset.updateAssetType";
    AssetTx["RemoveLocalMetadataKey"] = "asset.removeLocalMetadataKey";
    AssetTx["RemoveMetadataValue"] = "asset.removeMetadataValue";
    AssetTx["ExemptTickerAffirmation"] = "asset.exemptTickerAffirmation";
    AssetTx["RemoveTickerAffirmationExemption"] = "asset.removeTickerAffirmationExemption";
    AssetTx["PreApproveTicker"] = "asset.preApproveTicker";
    AssetTx["RemoveTickerPreApproval"] = "asset.removeTickerPreApproval";
})(AssetTx = exports.AssetTx || (exports.AssetTx = {}));
var CapitalDistributionTx;
(function (CapitalDistributionTx) {
    CapitalDistributionTx["Distribute"] = "capitalDistribution.distribute";
    CapitalDistributionTx["Claim"] = "capitalDistribution.claim";
    CapitalDistributionTx["PushBenefit"] = "capitalDistribution.pushBenefit";
    CapitalDistributionTx["Reclaim"] = "capitalDistribution.reclaim";
    CapitalDistributionTx["RemoveDistribution"] = "capitalDistribution.removeDistribution";
})(CapitalDistributionTx = exports.CapitalDistributionTx || (exports.CapitalDistributionTx = {}));
var CheckpointTx;
(function (CheckpointTx) {
    CheckpointTx["CreateCheckpoint"] = "checkpoint.createCheckpoint";
    CheckpointTx["SetSchedulesMaxComplexity"] = "checkpoint.setSchedulesMaxComplexity";
    CheckpointTx["CreateSchedule"] = "checkpoint.createSchedule";
    CheckpointTx["RemoveSchedule"] = "checkpoint.removeSchedule";
})(CheckpointTx = exports.CheckpointTx || (exports.CheckpointTx = {}));
var ComplianceManagerTx;
(function (ComplianceManagerTx) {
    ComplianceManagerTx["AddComplianceRequirement"] = "complianceManager.addComplianceRequirement";
    ComplianceManagerTx["RemoveComplianceRequirement"] = "complianceManager.removeComplianceRequirement";
    ComplianceManagerTx["ReplaceAssetCompliance"] = "complianceManager.replaceAssetCompliance";
    ComplianceManagerTx["ResetAssetCompliance"] = "complianceManager.resetAssetCompliance";
    ComplianceManagerTx["PauseAssetCompliance"] = "complianceManager.pauseAssetCompliance";
    ComplianceManagerTx["ResumeAssetCompliance"] = "complianceManager.resumeAssetCompliance";
    ComplianceManagerTx["AddDefaultTrustedClaimIssuer"] = "complianceManager.addDefaultTrustedClaimIssuer";
    ComplianceManagerTx["RemoveDefaultTrustedClaimIssuer"] = "complianceManager.removeDefaultTrustedClaimIssuer";
    ComplianceManagerTx["ChangeComplianceRequirement"] = "complianceManager.changeComplianceRequirement";
})(ComplianceManagerTx = exports.ComplianceManagerTx || (exports.ComplianceManagerTx = {}));
var CorporateActionTx;
(function (CorporateActionTx) {
    CorporateActionTx["SetMaxDetailsLength"] = "corporateAction.setMaxDetailsLength";
    CorporateActionTx["SetDefaultTargets"] = "corporateAction.setDefaultTargets";
    CorporateActionTx["SetDefaultWithholdingTax"] = "corporateAction.setDefaultWithholdingTax";
    CorporateActionTx["SetDidWithholdingTax"] = "corporateAction.setDidWithholdingTax";
    CorporateActionTx["InitiateCorporateAction"] = "corporateAction.initiateCorporateAction";
    CorporateActionTx["LinkCaDoc"] = "corporateAction.linkCaDoc";
    CorporateActionTx["RemoveCa"] = "corporateAction.removeCa";
    CorporateActionTx["ChangeRecordDate"] = "corporateAction.changeRecordDate";
    CorporateActionTx["InitiateCorporateActionAndDistribute"] = "corporateAction.initiateCorporateActionAndDistribute";
})(CorporateActionTx = exports.CorporateActionTx || (exports.CorporateActionTx = {}));
var CorporateBallotTx;
(function (CorporateBallotTx) {
    CorporateBallotTx["AttachBallot"] = "corporateBallot.attachBallot";
    CorporateBallotTx["Vote"] = "corporateBallot.vote";
    CorporateBallotTx["ChangeEnd"] = "corporateBallot.changeEnd";
    CorporateBallotTx["ChangeMeta"] = "corporateBallot.changeMeta";
    CorporateBallotTx["ChangeRcv"] = "corporateBallot.changeRcv";
    CorporateBallotTx["RemoveBallot"] = "corporateBallot.removeBallot";
})(CorporateBallotTx = exports.CorporateBallotTx || (exports.CorporateBallotTx = {}));
var PipsTx;
(function (PipsTx) {
    PipsTx["SetPruneHistoricalPips"] = "pips.setPruneHistoricalPips";
    PipsTx["SetMinProposalDeposit"] = "pips.setMinProposalDeposit";
    PipsTx["SetDefaultEnactmentPeriod"] = "pips.setDefaultEnactmentPeriod";
    PipsTx["SetPendingPipExpiry"] = "pips.setPendingPipExpiry";
    PipsTx["SetMaxPipSkipCount"] = "pips.setMaxPipSkipCount";
    PipsTx["SetActivePipLimit"] = "pips.setActivePipLimit";
    PipsTx["Propose"] = "pips.propose";
    PipsTx["Vote"] = "pips.vote";
    PipsTx["ApproveCommitteeProposal"] = "pips.approveCommitteeProposal";
    PipsTx["RejectProposal"] = "pips.rejectProposal";
    PipsTx["PruneProposal"] = "pips.pruneProposal";
    PipsTx["RescheduleExecution"] = "pips.rescheduleExecution";
    PipsTx["ClearSnapshot"] = "pips.clearSnapshot";
    PipsTx["Snapshot"] = "pips.snapshot";
    PipsTx["EnactSnapshotResults"] = "pips.enactSnapshotResults";
    PipsTx["ExecuteScheduledPip"] = "pips.executeScheduledPip";
    PipsTx["ExpireScheduledPip"] = "pips.expireScheduledPip";
})(PipsTx = exports.PipsTx || (exports.PipsTx = {}));
var PortfolioTx;
(function (PortfolioTx) {
    PortfolioTx["CreatePortfolio"] = "portfolio.createPortfolio";
    PortfolioTx["DeletePortfolio"] = "portfolio.deletePortfolio";
    PortfolioTx["MovePortfolioFunds"] = "portfolio.movePortfolioFunds";
    PortfolioTx["RenamePortfolio"] = "portfolio.renamePortfolio";
    PortfolioTx["QuitPortfolioCustody"] = "portfolio.quitPortfolioCustody";
    PortfolioTx["AcceptPortfolioCustody"] = "portfolio.acceptPortfolioCustody";
    PortfolioTx["MovePortfolioFundsV2"] = "portfolio.movePortfolioFundsV2";
    PortfolioTx["PreApprovePortfolio"] = "portfolio.preApprovePortfolio";
    PortfolioTx["RemovePortfolioPreApproval"] = "portfolio.removePortfolioPreApproval";
})(PortfolioTx = exports.PortfolioTx || (exports.PortfolioTx = {}));
var ProtocolFeeTx;
(function (ProtocolFeeTx) {
    ProtocolFeeTx["ChangeCoefficient"] = "protocolFee.changeCoefficient";
    ProtocolFeeTx["ChangeBaseFee"] = "protocolFee.changeBaseFee";
})(ProtocolFeeTx = exports.ProtocolFeeTx || (exports.ProtocolFeeTx = {}));
var SchedulerTx;
(function (SchedulerTx) {
    SchedulerTx["Schedule"] = "scheduler.schedule";
    SchedulerTx["Cancel"] = "scheduler.cancel";
    SchedulerTx["ScheduleNamed"] = "scheduler.scheduleNamed";
    SchedulerTx["CancelNamed"] = "scheduler.cancelNamed";
    SchedulerTx["ScheduleAfter"] = "scheduler.scheduleAfter";
    SchedulerTx["ScheduleNamedAfter"] = "scheduler.scheduleNamedAfter";
})(SchedulerTx = exports.SchedulerTx || (exports.SchedulerTx = {}));
var SettlementTx;
(function (SettlementTx) {
    SettlementTx["CreateVenue"] = "settlement.createVenue";
    SettlementTx["UpdateVenueDetails"] = "settlement.updateVenueDetails";
    SettlementTx["UpdateVenueType"] = "settlement.updateVenueType";
    SettlementTx["AddInstruction"] = "settlement.addInstruction";
    SettlementTx["AddAndAffirmInstruction"] = "settlement.addAndAffirmInstruction";
    SettlementTx["AffirmInstruction"] = "settlement.affirmInstruction";
    SettlementTx["WithdrawAffirmation"] = "settlement.withdrawAffirmation";
    SettlementTx["RejectInstruction"] = "settlement.rejectInstruction";
    SettlementTx["AffirmWithReceipts"] = "settlement.affirmWithReceipts";
    SettlementTx["ClaimReceipt"] = "settlement.claimReceipt";
    SettlementTx["UnclaimReceipt"] = "settlement.unclaimReceipt";
    SettlementTx["SetVenueFiltering"] = "settlement.setVenueFiltering";
    SettlementTx["AllowVenues"] = "settlement.allowVenues";
    SettlementTx["DisallowVenues"] = "settlement.disallowVenues";
    SettlementTx["ChangeReceiptValidity"] = "settlement.changeReceiptValidity";
    SettlementTx["ExecuteScheduledInstruction"] = "settlement.executeScheduledInstruction";
    SettlementTx["RescheduleInstruction"] = "settlement.rescheduleInstruction";
    SettlementTx["UpdateVenueSigners"] = "settlement.updateVenueSigners";
    SettlementTx["AddInstructionWithMemo"] = "settlement.addInstructionWithMemo";
    SettlementTx["AddAndAffirmInstructionWithMemo"] = "settlement.addAndAffirmInstructionWithMemo";
    SettlementTx["ExecuteManualInstruction"] = "settlement.executeManualInstruction";
    SettlementTx["PlaceholderClaimReceipt"] = "settlement.placeholderClaimReceipt";
    SettlementTx["PlaceholderUnclaimReceipt"] = "settlement.placeholderUnclaimReceipt";
    SettlementTx["AddInstructionWithMemoV2"] = "settlement.addInstructionWithMemoV2";
    SettlementTx["AddAndAffirmInstructionWithMemoV2"] = "settlement.addAndAffirmInstructionWithMemoV2";
    SettlementTx["AffirmInstructionV2"] = "settlement.affirmInstructionV2";
    SettlementTx["WithdrawAffirmationV2"] = "settlement.withdrawAffirmationV2";
    SettlementTx["RejectInstructionV2"] = "settlement.rejectInstructionV2";
    SettlementTx["ExecuteScheduledInstructionV2"] = "settlement.executeScheduledInstructionV2";
    SettlementTx["PlaceholderAddInstruction"] = "settlement.placeholderAddInstruction";
    SettlementTx["PlaceholderAddAndAffirmInstruction"] = "settlement.placeholderAddAndAffirmInstruction";
    SettlementTx["PlaceholderAffirmInstruction"] = "settlement.placeholderAffirmInstruction";
    SettlementTx["PlaceholderWithdrawAffirmation"] = "settlement.placeholderWithdrawAffirmation";
    SettlementTx["PlaceholderRejectInstruction"] = "settlement.placeholderRejectInstruction";
    SettlementTx["PlaceholderAddInstructionWithMemo"] = "settlement.placeholderAddInstructionWithMemo";
    SettlementTx["PlaceholderAddAndAffirmInstructionWithMemo"] = "settlement.placeholderAddAndAffirmInstructionWithMemo";
    SettlementTx["ExecuteScheduledInstructionV3"] = "settlement.executeScheduledInstructionV3";
})(SettlementTx = exports.SettlementTx || (exports.SettlementTx = {}));
var StatisticsTx;
(function (StatisticsTx) {
    StatisticsTx["AddTransferManager"] = "statistics.addTransferManager";
    StatisticsTx["RemoveTransferManager"] = "statistics.removeTransferManager";
    StatisticsTx["AddExemptedEntities"] = "statistics.addExemptedEntities";
    StatisticsTx["RemoveExemptedEntities"] = "statistics.removeExemptedEntities";
    StatisticsTx["SetAssetTransferCompliance"] = "statistics.setAssetTransferCompliance";
    StatisticsTx["SetEntitiesExempt"] = "statistics.setEntitiesExempt";
    StatisticsTx["SetActiveAssetStats"] = "statistics.setActiveAssetStats";
    StatisticsTx["BatchUpdateAssetStats"] = "statistics.batchUpdateAssetStats";
})(StatisticsTx = exports.StatisticsTx || (exports.StatisticsTx = {}));
var StoTx;
(function (StoTx) {
    StoTx["CreateFundraiser"] = "sto.createFundraiser";
    StoTx["Invest"] = "sto.invest";
    StoTx["FreezeFundraiser"] = "sto.freezeFundraiser";
    StoTx["UnfreezeFundraiser"] = "sto.unfreezeFundraiser";
    StoTx["ModifyFundraiserWindow"] = "sto.modifyFundraiserWindow";
    StoTx["Stop"] = "sto.stop";
})(StoTx = exports.StoTx || (exports.StoTx = {}));
var TreasuryTx;
(function (TreasuryTx) {
    TreasuryTx["Disbursement"] = "treasury.disbursement";
    TreasuryTx["Reimbursement"] = "treasury.reimbursement";
})(TreasuryTx = exports.TreasuryTx || (exports.TreasuryTx = {}));
var UtilityTx;
(function (UtilityTx) {
    UtilityTx["Batch"] = "utility.batch";
    UtilityTx["BatchAtomic"] = "utility.batchAtomic";
    UtilityTx["BatchOptimistic"] = "utility.batchOptimistic";
    UtilityTx["RelayTx"] = "utility.relayTx";
    UtilityTx["BatchAll"] = "utility.batchAll";
    UtilityTx["DispatchAs"] = "utility.dispatchAs";
    UtilityTx["ForceBatch"] = "utility.forceBatch";
    UtilityTx["WithWeight"] = "utility.withWeight";
    UtilityTx["BatchOld"] = "utility.batchOld";
})(UtilityTx = exports.UtilityTx || (exports.UtilityTx = {}));
var ExternalAgentsTx;
(function (ExternalAgentsTx) {
    ExternalAgentsTx["CreateGroup"] = "externalAgents.createGroup";
    ExternalAgentsTx["SetGroupPermissions"] = "externalAgents.setGroupPermissions";
    ExternalAgentsTx["RemoveAgent"] = "externalAgents.removeAgent";
    ExternalAgentsTx["Abdicate"] = "externalAgents.abdicate";
    ExternalAgentsTx["ChangeGroup"] = "externalAgents.changeGroup";
    ExternalAgentsTx["AcceptBecomeAgent"] = "externalAgents.acceptBecomeAgent";
    ExternalAgentsTx["CreateGroupAndAddAuth"] = "externalAgents.createGroupAndAddAuth";
    ExternalAgentsTx["CreateAndChangeCustomGroup"] = "externalAgents.createAndChangeCustomGroup";
})(ExternalAgentsTx = exports.ExternalAgentsTx || (exports.ExternalAgentsTx = {}));
var RelayerTx;
(function (RelayerTx) {
    RelayerTx["SetPayingKey"] = "relayer.setPayingKey";
    RelayerTx["AcceptPayingKey"] = "relayer.acceptPayingKey";
    RelayerTx["RemovePayingKey"] = "relayer.removePayingKey";
    RelayerTx["UpdatePolyxLimit"] = "relayer.updatePolyxLimit";
    RelayerTx["IncreasePolyxLimit"] = "relayer.increasePolyxLimit";
    RelayerTx["DecreasePolyxLimit"] = "relayer.decreasePolyxLimit";
})(RelayerTx = exports.RelayerTx || (exports.RelayerTx = {}));
var RewardsTx;
(function (RewardsTx) {
    RewardsTx["ClaimItnReward"] = "rewards.claimItnReward";
    RewardsTx["SetItnRewardStatus"] = "rewards.setItnRewardStatus";
})(RewardsTx = exports.RewardsTx || (exports.RewardsTx = {}));
var TestUtilsTx;
(function (TestUtilsTx) {
    TestUtilsTx["RegisterDid"] = "testUtils.registerDid";
    TestUtilsTx["MockCddRegisterDid"] = "testUtils.mockCddRegisterDid";
    TestUtilsTx["GetMyDid"] = "testUtils.getMyDid";
    TestUtilsTx["GetCddOf"] = "testUtils.getCddOf";
})(TestUtilsTx = exports.TestUtilsTx || (exports.TestUtilsTx = {}));
var PolymeshContractsTx;
(function (PolymeshContractsTx) {
    PolymeshContractsTx["Call"] = "polymeshContracts.call";
    PolymeshContractsTx["InstantiateWithCode"] = "polymeshContracts.instantiateWithCode";
    PolymeshContractsTx["Instantiate"] = "polymeshContracts.instantiate";
    PolymeshContractsTx["UploadCode"] = "polymeshContracts.uploadCode";
    PolymeshContractsTx["RemoveCode"] = "polymeshContracts.removeCode";
    PolymeshContractsTx["InstantiateWithCodePerms"] = "polymeshContracts.instantiateWithCodePerms";
    PolymeshContractsTx["InstantiateWithHashPerms"] = "polymeshContracts.instantiateWithHashPerms";
    PolymeshContractsTx["UpdateCallRuntimeWhitelist"] = "polymeshContracts.updateCallRuntimeWhitelist";
})(PolymeshContractsTx = exports.PolymeshContractsTx || (exports.PolymeshContractsTx = {}));
var PreimageTx;
(function (PreimageTx) {
    PreimageTx["NotePreimage"] = "preimage.notePreimage";
    PreimageTx["UnnotePreimage"] = "preimage.unnotePreimage";
    PreimageTx["RequestPreimage"] = "preimage.requestPreimage";
    PreimageTx["UnrequestPreimage"] = "preimage.unrequestPreimage";
})(PreimageTx = exports.PreimageTx || (exports.PreimageTx = {}));
var ContractsTx;
(function (ContractsTx) {
    ContractsTx["Call"] = "contracts.call";
    ContractsTx["InstantiateWithCode"] = "contracts.instantiateWithCode";
    ContractsTx["Instantiate"] = "contracts.instantiate";
    ContractsTx["UploadCode"] = "contracts.uploadCode";
    ContractsTx["RemoveCode"] = "contracts.removeCode";
    ContractsTx["CallOldWeight"] = "contracts.callOldWeight";
    ContractsTx["InstantiateWithCodeOldWeight"] = "contracts.instantiateWithCodeOldWeight";
    ContractsTx["InstantiateOldWeight"] = "contracts.instantiateOldWeight";
    ContractsTx["SetCode"] = "contracts.setCode";
})(ContractsTx = exports.ContractsTx || (exports.ContractsTx = {}));
var NftTx;
(function (NftTx) {
    NftTx["CreateNftCollection"] = "nft.createNftCollection";
    NftTx["IssueNft"] = "nft.issueNft";
    NftTx["RedeemNft"] = "nft.redeemNft";
})(NftTx = exports.NftTx || (exports.NftTx = {}));
var ModuleName;
(function (ModuleName) {
    ModuleName["System"] = "system";
    ModuleName["Babe"] = "babe";
    ModuleName["Timestamp"] = "timestamp";
    ModuleName["Indices"] = "indices";
    ModuleName["Authorship"] = "authorship";
    ModuleName["Balances"] = "balances";
    ModuleName["Identity"] = "identity";
    ModuleName["CddServiceProviders"] = "cddServiceProviders";
    ModuleName["PolymeshCommittee"] = "polymeshCommittee";
    ModuleName["CommitteeMembership"] = "committeeMembership";
    ModuleName["TechnicalCommittee"] = "technicalCommittee";
    ModuleName["TechnicalCommitteeMembership"] = "technicalCommitteeMembership";
    ModuleName["UpgradeCommittee"] = "upgradeCommittee";
    ModuleName["UpgradeCommitteeMembership"] = "upgradeCommitteeMembership";
    ModuleName["MultiSig"] = "multiSig";
    ModuleName["Bridge"] = "bridge";
    ModuleName["Staking"] = "staking";
    ModuleName["Session"] = "session";
    ModuleName["Grandpa"] = "grandpa";
    ModuleName["ImOnline"] = "imOnline";
    ModuleName["Sudo"] = "sudo";
    ModuleName["Asset"] = "asset";
    ModuleName["CapitalDistribution"] = "capitalDistribution";
    ModuleName["Checkpoint"] = "checkpoint";
    ModuleName["ComplianceManager"] = "complianceManager";
    ModuleName["CorporateAction"] = "corporateAction";
    ModuleName["CorporateBallot"] = "corporateBallot";
    ModuleName["Pips"] = "pips";
    ModuleName["Portfolio"] = "portfolio";
    ModuleName["ProtocolFee"] = "protocolFee";
    ModuleName["Scheduler"] = "scheduler";
    ModuleName["Settlement"] = "settlement";
    ModuleName["Statistics"] = "statistics";
    ModuleName["Sto"] = "sto";
    ModuleName["Treasury"] = "treasury";
    ModuleName["Utility"] = "utility";
    ModuleName["ExternalAgents"] = "externalAgents";
    ModuleName["Relayer"] = "relayer";
    ModuleName["Rewards"] = "rewards";
    ModuleName["TestUtils"] = "testUtils";
    ModuleName["PolymeshContracts"] = "polymeshContracts";
    ModuleName["Preimage"] = "preimage";
    ModuleName["Contracts"] = "contracts";
    ModuleName["Nft"] = "nft";
})(ModuleName = exports.ModuleName || (exports.ModuleName = {}));
// eslint-disable-next-line @typescript-eslint/naming-convention
exports.TxTags = {
    system: SystemTx,
    babe: BabeTx,
    timestamp: TimestampTx,
    indices: IndicesTx,
    authorship: AuthorshipTx,
    balances: BalancesTx,
    identity: IdentityTx,
    cddServiceProviders: CddServiceProvidersTx,
    polymeshCommittee: PolymeshCommitteeTx,
    committeeMembership: CommitteeMembershipTx,
    technicalCommittee: TechnicalCommitteeTx,
    technicalCommitteeMembership: TechnicalCommitteeMembershipTx,
    upgradeCommittee: UpgradeCommitteeTx,
    upgradeCommitteeMembership: UpgradeCommitteeMembershipTx,
    multiSig: MultiSigTx,
    bridge: BridgeTx,
    staking: StakingTx,
    session: SessionTx,
    grandpa: GrandpaTx,
    imOnline: ImOnlineTx,
    sudo: SudoTx,
    asset: AssetTx,
    capitalDistribution: CapitalDistributionTx,
    checkpoint: CheckpointTx,
    complianceManager: ComplianceManagerTx,
    corporateAction: CorporateActionTx,
    corporateBallot: CorporateBallotTx,
    pips: PipsTx,
    portfolio: PortfolioTx,
    protocolFee: ProtocolFeeTx,
    scheduler: SchedulerTx,
    settlement: SettlementTx,
    statistics: StatisticsTx,
    sto: StoTx,
    treasury: TreasuryTx,
    utility: UtilityTx,
    externalAgents: ExternalAgentsTx,
    relayer: RelayerTx,
    rewards: RewardsTx,
    testUtils: TestUtilsTx,
    polymeshContracts: PolymeshContractsTx,
    preimage: PreimageTx,
    contracts: ContractsTx,
    nft: NftTx,
};
