// @polymeshassociation/polymesh-sdk/middleware/enums.js
"use strict";
/* istanbul ignore file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.middlewareEnumMap = exports.ClaimTypeEnum = exports.AuthTypeEnum = exports.InstructionStatusEnum = exports.AuthorizationStatusEnum = exports.LegTypeEnum = exports.TransferRestrictionTypeEnum = exports.TransferComplianceTypeEnum = exports.StoStatus = exports.EventIdEnum = exports.ModuleIdEnum = exports.ProposalStateEnum = exports.BalanceTypeEnum = exports.CallIdEnum = exports.StatOpTypeEnum = exports.SettlementResultEnum = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "SettlementResultEnum", { enumerable: true, get: function () { return types_1.PublicEnum05F6462Eea; } });
Object.defineProperty(exports, "StatOpTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnum040F13614F; } });
Object.defineProperty(exports, "CallIdEnum", { enumerable: true, get: function () { return types_1.PublicEnum0Bf3C7D4Ef; } });
Object.defineProperty(exports, "BalanceTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnum5Df0F1D22C; } });
Object.defineProperty(exports, "ProposalStateEnum", { enumerable: true, get: function () { return types_1.PublicEnum6E6395Eda7; } });
Object.defineProperty(exports, "ModuleIdEnum", { enumerable: true, get: function () { return types_1.PublicEnum7A0B4Cc03E; } });
Object.defineProperty(exports, "EventIdEnum", { enumerable: true, get: function () { return types_1.PublicEnum8F5A39C8Ee; } });
Object.defineProperty(exports, "StoStatus", { enumerable: true, get: function () { return types_1.PublicEnum9Ceea16A44; } });
Object.defineProperty(exports, "TransferComplianceTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnum71Afee1504; } });
Object.defineProperty(exports, "TransferRestrictionTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnum72Fc336Bfe; } });
Object.defineProperty(exports, "LegTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnum2792Cc24Ad; } });
Object.defineProperty(exports, "AuthorizationStatusEnum", { enumerable: true, get: function () { return types_1.PublicEnum31117E090D; } });
Object.defineProperty(exports, "InstructionStatusEnum", { enumerable: true, get: function () { return types_1.PublicEnumB861Be9158; } });
Object.defineProperty(exports, "AuthTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnumCf84Da6Fd1; } });
Object.defineProperty(exports, "ClaimTypeEnum", { enumerable: true, get: function () { return types_1.PublicEnumEc3243D777; } });
exports.middlewareEnumMap = {
    SettlementResultEnum: 'PublicEnum05F6462Eea',
    StatOpTypeEnum: 'PublicEnum040F13614F',
    CallIdEnum: 'PublicEnum0Bf3C7D4Ef',
    BalanceTypeEnum: 'PublicEnum5Df0F1D22C',
    ProposalStateEnum: 'PublicEnum6E6395Eda7',
    ModuleIdEnum: 'PublicEnum7A0B4Cc03E',
    EventIdEnum: 'PublicEnum8F5A39C8Ee',
    StoStatus: 'PublicEnum9Ceea16A44',
    TransferComplianceTypeEnum: 'PublicEnum71Afee1504',
    TransferRestrictionTypeEnum: 'PublicEnum72Fc336Bfe',
    LegTypeEnum: 'PublicEnum2792Cc24Ad',
    AuthorizationStatusEnum: 'PublicEnum31117E090D',
    InstructionStatusEnum: 'PublicEnumB861Be9158',
    AuthTypeEnum: 'PublicEnumCf84Da6Fd1',
    ClaimTypeEnum: 'PublicEnumEc3243D777',
};
