// @polymeshassociation/polymesh-sdk/api/client/Identities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Identities = void 0;
const Venue_1 = require("../entities/Venue");
const internal_1 = require("../../internal");
const internal_2 = require("../../utils/internal");
/**
 * Handles all Identity related functionality
 */
class Identities {
    /**
     * @hidden
     */
    constructor(context) {
        this.context = context;
        this.registerIdentity = (0, internal_2.createProcedureMethod)({ getProcedureAndArgs: args => [internal_1.registerIdentity, args] }, context);
        this.attestPrimaryKeyRotation = (0, internal_2.createProcedureMethod)({ getProcedureAndArgs: args => [internal_1.attestPrimaryKeyRotation, args] }, context);
        this.rotatePrimaryKey = (0, internal_2.createProcedureMethod)({ getProcedureAndArgs: args => [internal_1.rotatePrimaryKey, args] }, context);
        this.createPortfolio = (0, internal_2.createProcedureMethod)({
            getProcedureAndArgs: args => [
                internal_1.createPortfolios,
                {
                    names: [args.name],
                },
            ],
            transformer: Venue_1.createPortfolioTransformer,
        }, context);
        this.createPortfolios = (0, internal_2.createProcedureMethod)({
            getProcedureAndArgs: args => [internal_1.createPortfolios, args],
        }, context);
        this.createChild = (0, internal_2.createProcedureMethod)({
            getProcedureAndArgs: args => [internal_1.createChildIdentity, args],
        }, context);
    }
    /**
     * Create an Identity instance from a DID
     *
     * @throws if there is no Identity with the passed DID
     */
    getIdentity(args) {
        return this.context.getIdentity(args.did);
    }
    /**
     * Create a ChildIdentity instance from a DID
     *
     * @throws if there is no ChildIdentity with the passed DID
     */
    getChildIdentity(args) {
        return this.context.getChildIdentity(args.did);
    }
    /**
     * Return whether the supplied Identity/DID exists
     */
    isIdentityValid(args) {
        return (0, internal_2.asIdentity)(args.identity, this.context).exists();
    }
}
exports.Identities = Identities;
